File: Projects/FlutterPerfLab/lib/modules/listtest/list_test_screen.dart
import 'dart:async';
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';
import 'package:go_router/go_router.dart';

import 'package:flutterperflab/models/product_data.dart';
import 'list_test_view_model.dart'; // Ensure ScrollCommand is defined here

/// A virtualized scroll UI engineered to evaluate hardware rendering throughput
/// and GPU frame-time during high-frequency programmatic scrolling.
class ListTestScreen extends StatelessWidget {
  const ListTestScreen({super.key});

  @override
  Widget build(BuildContext context) {
    return ChangeNotifierProvider(
      create: (_) => ListTestViewModel(),
      child: const _ListTestContent(),
    );
  }
}

class _ListTestContent extends StatefulWidget {
  const _ListTestContent();

  @override
  State<_ListTestContent> createState() => _ListTestContentState();
}

class _ListTestContentState extends State<_ListTestContent> {
  final ScrollController _scrollController = ScrollController();
  StreamSubscription<ScrollCommand>? _scrollSubscription;

  @override
  void initState() {
    super.initState();

    // Listens strictly to precise animation commands (index, duration, easing curve)
    // bypassing full widget tree recomposition.
    WidgetsBinding.instance.addPostFrameCallback((_) {
      final vm = context.read<ListTestViewModel>();

      _scrollSubscription = vm.scrollCommandStream.listen((command) {
        if (!_scrollController.hasClients) return;

        // Calculates the precise offset based on strict layout dimensions
        // instead of manual offset math to ensure deterministic scrolling.
        // Row calculation: 2 items per row.
        final int rowIndex = command.targetIndex ~/ 2;

        // 280 (Card Height) + 15 (Spacing).
        // Adding 16 for top padding offset.
        final double targetOffset = (rowIndex * 295.0) + 16.0;

        // Prevents out-of-bounds scrolling exceptions.
        final maxScroll = _scrollController.position.maxScrollExtent;
        final finalOffset = targetOffset > maxScroll ? maxScroll : targetOffset;

        _scrollController.animateTo(
          finalOffset,
          duration: command.duration,
          curve: command.curve,
        );
      });
    });
  }

  @override
  void dispose() {
    _scrollSubscription?.cancel();
    _scrollController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return PopScope(
      canPop: false,
      onPopInvokedWithResult: (bool didPop, dynamic result) {
        if (didPop) return;
        final vm = context.read<ListTestViewModel>();
        if (vm.isRunning) vm.stopTest(isFinished: false);
        context.go('/');
      },
      child: Scaffold(
        backgroundColor: const Color(0xFFF2F2F7),
        appBar: AppBar(
          title: const Text(
            "Storefront Performance",
            style: TextStyle(fontWeight: FontWeight.bold),
          ),
          backgroundColor: Colors.white,
          leading: IconButton(
            icon: const Icon(Icons.arrow_back),
            onPressed: () {
              final vm = context.read<ListTestViewModel>();
              if (vm.isRunning) vm.stopTest(isFinished: false);
              context.go('/');
            },
          ),
        ),
        body: Column(
          children: [
            // --- Telemetry Monitor Bar ---
            // Isolated Consumer prevents main view re-rendering.
            Consumer<ListTestViewModel>(
              builder: (context, vm, _) => Container(
                color: const Color(0xFFF2F2F7),
                padding: const EdgeInsets.symmetric(
                  horizontal: 16.0,
                  vertical: 12.0,
                ),
                child: Row(
                  children: [
                    Expanded(
                      child: Text(
                        vm.status,
                        style: const TextStyle(
                          fontSize: 12,
                          fontWeight: FontWeight.bold,
                          color: Colors.grey,
                        ),
                      ),
                    ),
                    if (vm.isLoadingMore)
                      const SizedBox(
                        width: 16,
                        height: 16,
                        child: CircularProgressIndicator(strokeWidth: 2),
                      ),
                  ],
                ),
              ),
            ),

            // --- Main Rendering Surface ---
            Expanded(
              // Using Selector ensures the heavy GridView is ONLY rebuilt when the
              // total count of products changes, not when status/text changes.
              child: Selector<ListTestViewModel, int>(
                selector: (context, vm) => vm.products.length,
                builder: (context, productCount, _) {
                  final vm = context.read<ListTestViewModel>();

                  return GridView.builder(
                    controller: _scrollController,
                    padding: const EdgeInsets.all(16.0),
                    gridDelegate:
                        const SliverGridDelegateWithFixedCrossAxisCount(
                          crossAxisCount: 2,
                          crossAxisSpacing: 15.0,
                          mainAxisSpacing: 15.0,
                          mainAxisExtent: 280.0,
                        ),
                    itemCount:
                        productCount +
                        (vm.isRunning && productCount < 300 ? 2 : 0),
                    itemBuilder: (context, index) {
                      if (index >= productCount) {
                        return const Center(child: CircularProgressIndicator());
                      }

                      final product = vm.products[index];
                      return RepaintBoundary(
                        key: ValueKey(product.id),
                        child: _ProductCardView(product: product),
                      );
                    },
                  );
                },
              ),
            ),

            // --- Controller Interface ---
            // Isolated Consumer for button states.
            Consumer<ListTestViewModel>(
              builder: (context, vm, _) => Container(
                color: Colors.white,
                padding: const EdgeInsets.all(16.0),
                child: Column(
                  children: [
                    if (!vm.isRunning && !vm.isReportReady)
                      SizedBox(
                        width: double.infinity,
                        height: 56,
                        child: ElevatedButton.icon(
                          style: ElevatedButton.styleFrom(
                            backgroundColor: Colors.blue,
                            foregroundColor: Colors.white,
                            shape: RoundedRectangleBorder(
                              borderRadius: BorderRadius.circular(12),
                            ),
                          ),
                          onPressed: () => vm.startTest(),
                          icon: const Icon(Icons.shopping_cart),
                          label: const Text(
                            "Start E-Commerce Benchmark",
                            style: TextStyle(fontWeight: FontWeight.bold),
                          ),
                        ),
                      )
                    else if (vm.isRunning)
                      SizedBox(
                        width: double.infinity,
                        height: 56,
                        child: ElevatedButton.icon(
                          style: ElevatedButton.styleFrom(
                            backgroundColor: Colors.red,
                            foregroundColor: Colors.white,
                            shape: RoundedRectangleBorder(
                              borderRadius: BorderRadius.circular(12),
                            ),
                          ),
                          onPressed: () => vm.stopTest(isFinished: false),
                          icon: const Icon(Icons.stop),
                          label: const Text(
                            "Stop Benchmark",
                            style: TextStyle(fontWeight: FontWeight.bold),
                          ),
                        ),
                      ),

                    if (vm.isReportReady && !vm.isRunning)
                      SizedBox(
                        width: double.infinity,
                        height: 56,
                        child: ElevatedButton.icon(
                          style: ElevatedButton.styleFrom(
                            backgroundColor: Colors.green,
                            foregroundColor: Colors.white,
                            shape: RoundedRectangleBorder(
                              borderRadius: BorderRadius.circular(12),
                            ),
                          ),
                          onPressed: () => vm.exportResults(),
                          icon: const Icon(Icons.system_update_alt),
                          label: const Text(
                            "Export Results (CSV)",
                            style: TextStyle(fontWeight: FontWeight.bold),
                          ),
                        ),
                      ),
                  ],
                ),
              ),
            ),
          ],
        ),
      ),
    );
  }
}

// MARK: - Product Card Component

const Map<String, String> _assetRegistry = {
  "laptop": "assets/laptop.jpg",
  "shoe": "assets/shoe.jpg",
  "smartphone": "assets/smartphone.jpg",
  "tablet": "assets/tablet.jpg",
  "watch": "assets/watch.jpg",
  "testphoto": "assets/testphoto.jpg",
};

class _ProductCardView extends StatelessWidget {
  final Product product;

  const _ProductCardView({required this.product});

  @override
  Widget build(BuildContext context) {
    final String cleanName = product.data.imageName.split('.').first;
    final String? assetPath = _assetRegistry[cleanName];

    return Container(
      decoration: BoxDecoration(
        color: Colors.white,
        borderRadius: BorderRadius.circular(12.0),
        boxShadow: [
          BoxShadow(
            color: Colors.black.withOpacity(0.1),
            blurRadius: 5,
            offset: const Offset(0, 2),
          ),
        ],
      ),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          SizedBox(
            height: 150,
            child: Stack(
              children: [
                ClipRRect(
                  borderRadius: const BorderRadius.vertical(
                    top: Radius.circular(12.0),
                  ),
                  child: Container(
                    width: double.infinity,
                    color: Colors.white,
                    child: assetPath != null
                        ? Image.asset(
                            assetPath,
                            fit: BoxFit.contain,
                            // Reduces anti-aliasing CPU cost.
                            filterQuality: FilterQuality.low,
                            // Prevents flickering.
                            gaplessPlayback: true,
                          )
                        : Container(
                            color: Colors.grey.shade300,
                            child: const Center(
                              child: Text(
                                "Asset Missing",
                                style: TextStyle(
                                  fontSize: 10,
                                  color: Colors.grey,
                                ),
                              ),
                            ),
                          ),
                  ),
                ),
                Positioned(
                  top: 8,
                  right: 8,
                  child: Container(
                    padding: const EdgeInsets.symmetric(
                      horizontal: 8,
                      vertical: 4,
                    ),
                    decoration: BoxDecoration(
                      color: Colors.red,
                      borderRadius: BorderRadius.circular(4),
                    ),
                    child: Text(
                      product.data.discount,
                      style: const TextStyle(
                        fontSize: 10,
                        fontWeight: FontWeight.bold,
                        color: Colors.white,
                      ),
                    ),
                  ),
                ),
              ],
            ),
          ),
          Expanded(
            child: Padding(
              padding: const EdgeInsets.all(10.0),
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  Text(
                    product.data.name,
                    maxLines: 2,
                    overflow: TextOverflow.ellipsis,
                    style: const TextStyle(
                      fontSize: 14,
                      fontWeight: FontWeight.w600,
                      color: Colors.black87,
                    ),
                  ),
                  const SizedBox(height: 4),
                  Text(
                    product.data.category,
                    style: const TextStyle(fontSize: 12, color: Colors.grey),
                  ),
                  const Spacer(),
                  Row(
                    crossAxisAlignment: CrossAxisAlignment.end,
                    children: [
                      Column(
                        crossAxisAlignment: CrossAxisAlignment.start,
                        children: [
                          Text(
                            product.data.oldPrice,
                            style: const TextStyle(
                              fontSize: 10,
                              color: Colors.grey,
                              decoration: TextDecoration.lineThrough,
                            ),
                          ),
                          Text(
                            product.data.price,
                            style: const TextStyle(
                              fontSize: 16,
                              fontWeight: FontWeight.bold,
                              color: Colors.orange,
                            ),
                          ),
                        ],
                      ),
                      const Spacer(),
                      const Icon(
                        Icons.add_circle,
                        color: Colors.blue,
                        size: 24,
                      ),
                    ],
                  ),
                ],
              ),
            ),
          ),
        ],
      ),
    );
  }
}

File: Projects/NativeKotlinPerfLab/app/src/main/java/com/tez/nativekotlinperflabapp/ui/features/listtest/ListTestScreen.kt
package com.tez.nativekotlinperflabapp.ui.features.listtest

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.grid.*
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextDecoration
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import androidx.lifecycle.viewmodel.compose.viewModel
import coil3.compose.AsyncImage
import com.tez.nativekotlinperflabapp.models.Product

/**
 * The presentation layer for evaluating high-frequency list rendering and scrolling performance.
 */
@Composable
fun ListTestScreen(
    viewModel: ListTestViewModel = viewModel()
) {
    val context = LocalContext.current

    // Uses direct state access to prevent unnecessary collection allocations during updates.
    val products = viewModel.products

    // Lifecycle-aware state observations
    val isRunning by viewModel.isRunning.collectAsStateWithLifecycle()
    val isLoadingMore by viewModel.isLoadingMore.collectAsStateWithLifecycle()
    val status by viewModel.status.collectAsStateWithLifecycle()
    val exportUri by viewModel.exportUri.collectAsStateWithLifecycle()

    val gridState = rememberLazyGridState()

    // Encapsulates automated scrolling commands from the ViewModel to the UI thread,
    // simulating rapid, deterministic user interaction.
    LaunchedEffect(Unit) {
        viewModel.scrollCommand.collect { command ->
            if (command.targetIndex in products.indices) {
                gridState.animateScrollToItem(command.targetIndex)
            }
        }
    }

    DisposableEffect(Unit) {
        onDispose { viewModel.stopTest(isFinished = false) }
    }

    Column(modifier = Modifier.fillMaxSize()) {

        // --- Status Bar ---
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .background(Color(0xFFF2F2F7))
                .padding(horizontal = 16.dp, vertical = 12.dp),
            verticalAlignment = Alignment.CenterVertically
        ) {
            Text(
                text = status,
                fontSize = 12.sp,
                fontWeight = FontWeight.Bold,
                color = Color.Gray,
                modifier = Modifier.weight(1f)
            )

            if (isLoadingMore) {
                CircularProgressIndicator(modifier = Modifier.size(16.dp), strokeWidth = 2.dp)
            }
        }

        // --- Main Rendering Surface ---
        Box(
            modifier = Modifier
                .weight(1f)
                .background(Color(0xFFF2F2F7))
        ) {
            LazyVerticalGrid(
                columns = GridCells.Fixed(2),
                state = gridState,
                contentPadding = PaddingValues(16.dp),
                horizontalArrangement = Arrangement.spacedBy(15.dp),
                verticalArrangement = Arrangement.spacedBy(15.dp),
                modifier = Modifier.fillMaxSize()
            ) {
                // Using explicit keys bypasses heavy hash-map calculations in Compose,
                // capturing pure rendering performance.
                items(
                    items = products,
                    key = { it.id },
                    contentType = { "product" }
                ) { product ->
                    ProductCardView(product = product)
                }

                if (isRunning && products.size < 300) {
                    item(span = { GridItemSpan(2) }) {
                        Box(
                            modifier = Modifier.fillMaxWidth().padding(16.dp),
                            contentAlignment = Alignment.Center
                        ) {
                            CircularProgressIndicator()
                        }
                    }
                }
            }
        }

        // --- Controller Interface ---
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .background(MaterialTheme.colorScheme.surface)
                .padding(16.dp),
            verticalArrangement = Arrangement.spacedBy(12.dp)
        ) {
            if (!isRunning) {
                Button(
                    onClick = { viewModel.startTest() },
                    modifier = Modifier.fillMaxWidth().height(56.dp),
                    colors = ButtonDefaults.buttonColors(containerColor = Color(0xFF007AFF)),
                    shape = RoundedCornerShape(12.dp)
                ) {
                    Icon(Icons.Default.ShoppingCart, contentDescription = null, modifier = Modifier.size(18.dp))
                    Spacer(modifier = Modifier.width(8.dp))
                    Text("Start E-Commerce Simulation", fontWeight = FontWeight.Bold)
                }
            } else {
                Button(
                    onClick = { viewModel.stopTest(isFinished = false) },
                    modifier = Modifier.fillMaxWidth().height(56.dp),
                    colors = ButtonDefaults.buttonColors(containerColor = Color.Red),
                    shape = RoundedCornerShape(12.dp)
                ) {
                    Icon(Icons.Default.Stop, contentDescription = null, modifier = Modifier.size(18.dp))
                    Spacer(modifier = Modifier.width(8.dp))
                    Text("Stop Simulation", fontWeight = FontWeight.Bold)
                }
            }

            if (exportUri != null && !isRunning) {
                Button(
                    onClick = { viewModel.exportResults(context) },
                    modifier = Modifier.fillMaxWidth().height(56.dp),
                    colors = ButtonDefaults.buttonColors(containerColor = Color(0xFF34C759)),
                    shape = RoundedCornerShape(12.dp)
                ) {
                    Icon(Icons.Default.SystemUpdateAlt, contentDescription = null, modifier = Modifier.size(18.dp))
                    Spacer(modifier = Modifier.width(8.dp))
                    Text("Export Results (CSV)", fontWeight = FontWeight.Bold)
                }
            }
        }
    }
}

// --- ASSET REGISTRY ---
private val ASSET_REGISTRY = mapOf(
    "laptop"     to "laptop.jpg",
    "shoe"       to "shoe.jpg",
    "smartphone" to "smartphone.jpg",
    "tablet"     to "tablet.jpg",
    "watch"      to "watch.jpg",
    "testphoto"  to "testphoto.jpg"
)

@Composable
fun ProductCardView(product: Product) {

    // Resolves asset path strings through a centralized registry to ensure fast lookups
    // during rapid scrolling cycles.
    val assetPath = remember(product.data.imageName) {
        val cleanName = product.data.imageName.substringBeforeLast(".")
        val fileName = ASSET_REGISTRY[cleanName]
        if (fileName != null) "file:///android_asset/$fileName" else null
    }

    Card(
        modifier = Modifier.fillMaxWidth().height(280.dp),
        shape = RoundedCornerShape(12.dp),
        elevation = CardDefaults.cardElevation(defaultElevation = 5.dp),
        colors = CardDefaults.cardColors(containerColor = MaterialTheme.colorScheme.surface)
    ) {
        Column {
            Box(
                modifier = Modifier
                    .fillMaxWidth()
                    .height(150.dp)
                    .clip(RoundedCornerShape(topStart = 12.dp, topEnd = 12.dp))
                    .background(Color.White)
            ) {
                if (assetPath != null) {
                    // Uses Coil's internal bitmap pooling to prevent UI thread starvation
                    // during high-velocity recycling.
                    AsyncImage(
                        model = assetPath,
                        contentDescription = product.data.name,
                        contentScale = ContentScale.Fit,
                        modifier = Modifier.fillMaxSize().padding(10.dp)
                    )
                } else {
                    Box(modifier = Modifier.fillMaxSize().background(Color.LightGray), contentAlignment = Alignment.Center) {
                        Text("Asset Missing", fontSize = 10.sp, color = Color.Gray)
                    }
                }

                Text(
                    text = product.data.discount,
                    fontSize = 10.sp,
                    fontWeight = FontWeight.Bold,
                    color = Color.White,
                    modifier = Modifier
                        .align(Alignment.TopEnd)
                        .padding(8.dp)
                        .background(Color.Red, RoundedCornerShape(4.dp))
                        .padding(horizontal = 8.dp, vertical = 4.dp)
                )
            }

            // --- PRODUCT METADATA ---
            Column(
                modifier = Modifier.fillMaxSize().padding(10.dp),
                verticalArrangement = Arrangement.SpaceBetween
            ) {
                Column(verticalArrangement = Arrangement.spacedBy(6.dp)) {
                    Text(
                        text = product.data.name,
                        fontSize = 14.sp,
                        fontWeight = FontWeight.SemiBold,
                        color = MaterialTheme.colorScheme.onSurface,
                        maxLines = 2,
                        overflow = TextOverflow.Ellipsis,
                        modifier = Modifier.heightIn(min = 40.dp)
                    )
                    Text(text = product.data.category, fontSize = 12.sp, color = Color.Gray)
                }

                Row(
                    modifier = Modifier.fillMaxWidth(),
                    horizontalArrangement = Arrangement.SpaceBetween,
                    verticalAlignment = Alignment.Bottom
                ) {
                    Column(verticalArrangement = Arrangement.spacedBy(2.dp)) {
                        Text(
                            text = product.data.oldPrice,
                            fontSize = 10.sp,
                            color = Color.Gray,
                            textDecoration = TextDecoration.LineThrough
                        )
                        Text(
                            text = product.data.price,
                            fontSize = 16.sp,
                            fontWeight = FontWeight.Bold,
                            color = Color(0xFFFF9500)
                        )
                    }
                    Icon(
                        imageVector = Icons.Default.AddCircle,
                        contentDescription = "Cart Trigger",
                        tint = Color(0xFF007AFF),
                        modifier = Modifier.size(24.dp)
                    )
                }
            }
        }
    }
}